File: backend-legacy/services/embeddings_service.py
import os
import time
import google.generativeai as genai
from dotenv import load_dotenv

load_dotenv()

_GEMINI_API_KEY = os.getenv("GEMINI_API_KEY")
_EMBEDDING_MODEL = os.getenv("EMBEDDING_MODEL", "models/embedding-001")

if not _GEMINI_API_KEY:
    raise ValueError("GEMINI_API_KEY is not set in .env file")

genai.configure(api_key=_GEMINI_API_KEY)


def embed_text(text: str) -> list[float]:
    for attempt in range(3):
        try:
            result = genai.embed_content(
                model=_EMBEDDING_MODEL,
                content=text,
                task_type="retrieval_document"
            )
            return result["embedding"]
        except Exception as e:
            if attempt < 2:
                time.sleep(1)
            else:
                raise ValueError(f"Embedding failed after 3 attempts: {e}")
    return []


def embed_query(text: str) -> list[float]:
    for attempt in range(3):
        try:
            result = genai.embed_content(
                model=_EMBEDDING_MODEL,
                content=text,
                task_type="retrieval_query"
            )
            return result["embedding"]
        except Exception as e:
            if attempt < 2:
                time.sleep(1)
            else:
                raise ValueError(f"Query embedding failed after 3 attempts: {e}")
    return []

File: backend/src/embeddings/embeddings.service.ts
import { Injectable, OnModuleInit } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { GoogleGenerativeAI, TaskType } from '@google/generative-ai';

@Injectable()
export class EmbeddingsService implements OnModuleInit {
  private genAI: GoogleGenerativeAI;
  private embeddingModel: string;

  constructor(private configService: ConfigService) {}

  onModuleInit() {
    const apiKey = this.configService.get<string>('GEMINI_API_KEY');
    if (!apiKey) {
      throw new Error('GEMINI_API_KEY is not set');
    }
    this.genAI = new GoogleGenerativeAI(apiKey);
    this.embeddingModel = this.configService.get<string>('EMBEDDING_MODEL', 'models/embedding-001');
  }

  async embedText(text: string): Promise<number[]> {
    return this.embedWithRetry(text, TaskType.RETRIEVAL_DOCUMENT);
  }

  async embedQuery(text: string): Promise<number[]> {
    return this.embedWithRetry(text, TaskType.RETRIEVAL_QUERY);
  }

  private async embedWithRetry(text: string, taskType: TaskType, attempts = 3): Promise<number[]> {
    for (let i = 0; i < attempts; i++) {
      try {
        const model = this.genAI.getGenerativeModel({ model: this.embeddingModel });
        const result = await model.embedContent({
          content: { role: 'user', parts: [{ text }] },
          taskType,
        });
        return result.embedding.values;
      } catch (error) {
        if (i === attempts - 1) throw error;
        await new Promise((resolve) => setTimeout(resolve, 1000));
      }
    }
    return [];
  }
}
